ncm-spma/src/main/perl/spma/yumng.pm

```diff
@@ -0,0 +1,810 @@
+    return 0 if !defined($preinstalled_rpms);
+    $preinstalled_rpms =~ s/\(none\)/0/g;
+    my $preinstalled = Set::Scalar->new( split ( /\n/, $preinstalled_rpms ) );
+
+    if ( !$NoAction && $spma_run ) {
+	# Attempt to update YUM and plugins to latest available release.
+	my $updateyum = $self->execute_yum_command( ["yum update -y yum\* " . YUM_PLUGIN_OPTS], "updating YUM stack" );
+	if ($yum_warned) {
+	    # Be sure to use current version of YUM available in metadata (should handle downgrades).
+	    my $yum_reinstall = $self->execute_yum_command( ["yum reinstall -y yum\* " . YUM_PLUGIN_OPTS], "making sure we are using correct YUM" );
+	    # Expiration of cache is not enough sometimes.
+	    # https://bugzilla.redhat.com/show_bug.cgi?id=1151074
+	    my $yum_clean = $self->execute_yum_command( ["yum clean all " . YUM_PLUGIN_OPTS], "resetting hard YUM state" );
+	    return 0 if !defined($yum_clean);
+	}
+    }
+
+    # Clean up YUM state - worth to be thorough there
+    my $yum_clean = $self->execute_yum_command( ["yum clean expire-cache " . YUM_PLUGIN_OPTS], "expiring YUM caches", 1 );
+    if ($yum_warned) {
+        # Expiration of cache is not enough sometimes.
+        # https://bugzilla.redhat.com/show_bug.cgi?id=1151074
+        my $yum_clean = $self->execute_yum_command( ["yum clean all " . YUM_PLUGIN_OPTS], "resetting hard YUM state", 1 );
+        return 0 if !defined($yum_clean);
+    }
+
+    return 0 if !defined($yum_clean);
+    $yum_clean = $self->execute_yum_command( ["yum makecache " . YUM_PLUGIN_OPTS], "generating YUM cache", 1 );
+    return 0 if !defined($yum_clean);
+
+    my @files = glob "{/tmp/*.yumtx,/var/lib/yum/transaction*}";
+    foreach my $file (@files) {
+        if ( !unlink $file ) {
+            $self->info("unable to remove file $file: $!");
+        }
+    }
+    my @dirs = glob "/var/tmp/yum-root*";
+    foreach my $dir (@dirs) {
+        if ( !rmtree $dir) {
+            $self->info("unable to remove directory $dir: $!");
+        }
+    }
+
+    # Test whether comps/groups are sane.
+    {
+	$yum_warned = 0;
+	my $groupinfo = $self->execute_yum_command( [ "yum groupinfo core " . YUM_PLUGIN_OPTS ], "testing comps/groups sanity", 1 );
+	if ( $yum_warned ) {
+	    $self->error("Groups are not sane - core group missing. Will not continue.");
+	    return 0;
+	}
+    }
+
+    # Query metadata for version locked packages including Epoch and write versionlock.list
+    my $repodata_rpms = $self->execute_yum_command( [REPO_AVAIL_PKGS], "fetching full package list", 1 );
+    return 0 if !defined($repodata_rpms);
+
+    # Test whether locked packages are present in the metadata
+    my $repoquery_list = Set::Scalar->new( reverse(split ( /\n/, $repodata_rpms)) );
+    my $repoquery_list_noepoch = Set::Scalar->new();
+    my $locked_found           = Set::Scalar->new();
+    my $locked_found_noepoch   = Set::Scalar->new();
+    $self->info( "(" . $repoquery_list->size . " total packages)" );
```

Comment:
The '(' is used to signify a list of packages is going to be printed. This list might be long, hnece the ().